# Tugas Akhir
# Chihago Crimes Dataset

##### Anggota Kelompok:
1. Aditya Arya Hendrady (null)
2. Arfi Renaldi (null)
3. Harry Akbar Ali Munir (1506689080)
4. Rifki Adrian (1506726946)

###### Kelas A
###### Faculty of Computer Science, University of Indonesia

## 1) Environment Preparation

### 1.a) Library installation

In [ ]:
import pandas as pd
import numpy as np
import random
import seaborn as sns
import matplotlib.pyplot as plt

from sklearn import preprocessing
from sklearn.tree import DecisionTreeRegressor
from sklearn.cross_validation import train_test_split
from sklearn import metrics
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.preprocessing import LabelEncoder

## 2) Data Preparation

In [10]:
df = pd.read_csv("Crimes_-_2001_to_present.csv")

print("Jumlah row: {}".format(df.shape[0]))
print("Jumlah feature: {}".format(df.shape[1]))
df.head()

Jumlah feature: 6582553
Jumlah row: 22


ID Case Number                    Date                   Block  IUCR  \
0  10000092    HY189866  03/18/2015 07:44:00 PM         047XX W OHIO ST  041A   
1  10000094    HY190059  03/18/2015 11:00:00 PM  066XX S MARSHFIELD AVE  4625   
2  10000095    HY190052  03/18/2015 10:45:00 PM   044XX S LAKE PARK AVE  0486   
3  10000096    HY190054  03/18/2015 10:30:00 PM    051XX S MICHIGAN AVE  0460   
4  10000097    HY189976  03/18/2015 09:00:00 PM        047XX W ADAMS ST  031A   

    Primary Type              Description Location Description Arrest  \
0        BATTERY      AGGRAVATED: HANDGUN               STREET  False   
1  OTHER OFFENSE         PAROLE VIOLATION               STREET   True   
2        BATTERY  DOMESTIC BATTERY SIMPLE            APARTMENT  False   
3        BATTERY                   SIMPLE            APARTMENT  False   
4        ROBBERY           ARMED: HANDGUN             SIDEWALK  False   

  Domestic              ...                Ward  Community Area  FBI Code  \
0    False              ...                28.0            25.0       04B   
1    False              ...                15.0            67.0        26   
2     True              ...                 4.0            39.0       08B   
3    False              ...                 3.0            40.0       08B   
4    False              ...                28.0            25.0        03   

   X Coordinate Y Coordinate  Year              Updated On   Latitude  \
0     1144606.0    1903566.0  2015  02/10/2018 03:50:01 PM  41.891399   
1     1166468.0    1860715.0  2015  02/10/2018 03:50:01 PM  41.773372   
2     1185075.0    1875622.0  2015  02/10/2018 03:50:01 PM  41.813861   
3     1178033.0    1870804.0  2015  02/10/2018 03:50:01 PM  41.800802   
4     1144920.0    1898709.0  2015  02/10/2018 03:50:01 PM  41.878065   

   Longitude                       Location  
0 -87.744385  (41.891398861, -87.744384567)  
1 -87.665319  (41.773371528, -87.665319468)  
2 -87.596643   (41.81386068, -87.596642837)  
3 -87.622619  (41.800802415, -87.622619343)  
4 -87.743354  (41.878064761, -87.743354013)  

[5 rows x 22 columns]

### 2.a) Data Cleansing

#### Sampling

Sampling diperlukan karena terdapat jumlah baris yang sangat banyak (6582553 baris) sehingga akan menyebabkan proses komputasi yang besar. Data sample diambil sebanyak 30% sebanyak data populasi

In [14]:
import numpy as np
import random

df_sample = df.sample(frac=0.3)
print("Jumlah row: {}".format(df_sample.shape[0]))
print("Jumlah feature: {}".format(df_sample.shape[1]))
df_sample

Jumlah row: 197476
Jumlah feature: 22


ID Case Number                    Date  \
2769935   8029141    HT260262  04/21/2011 09:00:00 AM   
2686760   7831046    HS630168  11/23/2010 08:55:00 PM   
4428505   2974040    HJ641142  09/20/2003 12:35:00 AM   
6446774   7078714    HR486615  08/15/2009 01:00:00 PM   
5725236   5588831    HN393920  06/09/2007 04:40:15 AM   
5901840   5950766    HN750491  12/09/2007 04:00:00 PM   
5896649   5937946    HN736503  12/01/2007 10:25:08 AM   
3109922   8839850    HV512390  10/10/2012 02:55:00 PM   
1009667   3266004    HK228334  03/08/2004 10:00:00 PM   
4058347   2470840    HH801488  11/25/2002 07:15:00 AM   
1486583   4869467    HM473997  07/12/2006 03:30:00 PM   
1558222   5367939    HN215122  03/05/2007 11:00:00 AM   
2408209   7377440    HS179138  02/24/2010 01:00:00 AM   
3478717   9427187    HW570948  12/14/2013 06:00:00 PM   
4318963   2813274    HJ467389  07/02/2003 08:30:00 AM   
2340130   7136798    HR546056  09/19/2009 09:00:00 PM   
4960540   3858030    HL234780  03/14/2005 02:00:00 AM   
404105   10746739    HZ508042  11/08/2016 06:33:00 PM   
1094416   3710918    HK815196  12/18/2004 03:20:00 AM   
5269676   4563395    HM153333  01/28/2006 01:00:00 PM   
4510351   3090848    HJ813604  12/12/2003 12:05:00 PM   
5755743   5650929    HN461281  07/10/2007 05:00:00 PM   
2125358   1716183     G513016  08/27/2001 08:45:00 PM   
3594538   9705789    HX355552  07/22/2014 12:50:00 PM   
3456815   9391629    HW534876  11/15/2013 06:20:00 PM   
2412640   7384382    HS186519  02/23/2010 03:30:00 PM   
6193794   6571017    HP640128  10/16/2008 05:15:00 PM   
510740   10969135    JA292880  06/04/2017 08:00:00 PM   
1354206   4833944    HM445578  06/30/2006 04:13:42 AM   
934782    2978221    HJ673922  10/05/2003 04:55:00 PM   
...           ...         ...                     ...   
4063693   2479851    HH811283  12/01/2002 01:50:00 AM   
834147    2440404    HH757031  11/03/2002 08:45:00 PM   
4360996  10933344    JA249490  05/03/2017 03:50:00 PM   
1570003   5472865    HN297606  04/18/2007 08:00:00 PM   
5870609   5879483    HN683953  11/01/2007 09:35:00 PM   
3300812   9143105    HW288238  05/23/2013 11:00:00 AM   
1558306   5368793    HM577488  09/02/2006 01:22:00 PM   
5895666   5936363    HN735118  11/30/2007 02:55:00 PM   
3910331   2230000    HH502007  07/10/2002 10:50:00 PM   
6228161   6637384    HP682287  11/14/2008 07:30:00 AM   
5345517   4708979    HM311298  04/24/2006 09:40:00 AM   
5563303   5192289    HM783036  12/01/2006 11:25:00 AM   
3430292   9348259    HW491470  10/13/2013 02:54:00 PM   
46068    10086845    HY275480  05/26/2015 09:00:00 AM   
6271094   6722107    HR140149  01/26/2009 10:15:00 PM   
6033685   6246704    HP333296  05/13/2008 02:00:00 PM   
857072    2572797    HJ143417  01/23/2003 09:40:00 PM   
1848117   1364881     G071486  02/03/2001 08:30:00 PM   
697192   11204098    JB116266  01/14/2018 06:00:00 PM   
5108933   4216645    HL543122  08/11/2005 09:00:00 AM   
854890    2560182    HJ145160  01/24/2003 09:09:00 PM   
3352574   9223974    HW370323  07/19/2013 10:33:00 PM   
2036039   1604312     G371705  06/26/2001 09:04:51 AM   
2799     10005404    HY195195  03/23/2015 10:39:00 AM   
1170588   4134310    HL467860  07/07/2005 08:30:00 AM   
5481721   4999020    HM375354  05/26/2006 07:45:41 PM   
767637    6424271    HP508436  08/12/2008 12:00:00 AM   
3110183   8840138    HV513072  10/10/2012 09:47:00 PM   
1078753   3607141    HK700166  10/21/2004 10:00:00 AM   
4301430   2790146    HJ440750  06/19/2003 12:00:00 PM   

                            Block  IUCR         Primary Type  \
2769935         065XX S HONORE ST  0620             BURGLARY   
2686760      0000X N KILBOURN AVE  1525         PROSTITUTION   
4428505        009XX N CICERO AVE  1506         PROSTITUTION   
6446774      016XX N BOSWORTH AVE  0610             BURGLARY   
5725236       021XX S HARDING AVE  0486              BATTERY   
5901840        068XX S HALSTED ST  0820                THEFT   
5896649       012XX W HASTINGS S

#### Mengecek Null Values

In [16]:
df_sample.isnull().sum()

ID                          0
Case Number                 0
Date                        0
Block                       0
IUCR                        0
Primary Type                0
Description                 0
Location Description      110
Arrest                      0
Domestic                    0
Beat                        0
District                    2
Ward                    18388
Community Area          18423
FBI Code                    0
X Coordinate             3783
Y Coordinate             3783
Year                        0
Updated On                  0
Latitude                 3783
Longitude                3783
Location                 3783
dtype: int64

Dikarenakan adanya null values, sehingga kami perlu menghandle missing values.
Penanganan missing values ini dibagi menjadi 2 kategori yaitu pada data kategorikal dan data numerikal.
- Pada data numerikal, kami memutuskan untuk mengisi missing values dengan nilai median pada setiap fitur
- Pada data kategorikal, kami memutuskan untuk mengisi missing values dengan nilai modus pada setiap fitur

In [56]:
location_description_mode = df_sample["Location Description"].mode()
district_mean = df_sample["District"].mean()
ward_mean = df_sample["Ward"].mean()
community_area_mean = df_sample["Community Area"].mean()
x_coordinate_mean = df_sample["X Coordinate"].mean()
y_coordinate_mean = df_sample["Y Coordinate"].mean()
latitude_mean = df_sample["Latitude"].mean()
longitude_mean = df_sample["Longitude"].mean()

df_clean = df_sample.fillna(value={
    "Location Description": location_description_mode[0],
    "District" : district_mean,
    "Ward": ward_mean,
    "Community Area": community_area_mean,
    "X Coordinate": x_coordinate_mean,
    "Y Coordinate" : y_coordinate_mean,
    "Latitude" : latitude_mean,
    "Longitude" : longitude_mean,
    "Location" : "({},{})".format(latitude_mean, longitude_mean)
})

df_clean.isnull().sum()

ID                      0
Case Number             0
Date                    0
Block                   0
IUCR                    0
Primary Type            0
Description             0
Location Description    0
Arrest                  0
Domestic                0
Beat                    0
District                0
Ward                    0
Community Area          0
FBI Code                0
X Coordinate            0
Y Coordinate            0
Year                    0
Updated On              0
Latitude                0
Longitude               0
Location                0
dtype: int64

####  Mengecek Null Values
Sebelum data digunakan lebih 

### 2.b) Feature Extraction

### 2.c) Data Analysis

## 3) Accuracy Testing

### 3.a) Test a few machine learning algorithm with your dataset